### PR TITLE
Preliminary PR for optim transforms and model

### PR DESCRIPTION
**Pull request by ProGamerGov**
This preliminary PR should make it easier for @greentfrapp to start testing with the Inception V1 model. It also removes some of the unnecessary imports, and implements changes as per our previous discussions.

Changes made:

* Removed Kornia imports.

* Removed ImageIO import.

* Replaced some NumPy code with PyTorch code.

* Removed Matplotlib code from image saving function.

* Uploaded new model to replace the old one. The new model has RGB to BGR built into it's transform inputs option, so we'll probably have to figure out normalization and RGB to BGR conversions outside of the model in the future.

* Added tutorial notebook for the InceptionV1 / GoogleNet / Inception5h model, based on the Torchvision notebook.

* Added DeepDream objective.

* Added L1 objective. Probably doesn't work correctly as the current objective system always gets the negative mean of the loss function. The most common use case would be using the L1 objective as a penalty.

* Added L2 objective.

* Added Total Variation Objective.

* Added improved Diversity objective. There's now no need for `F.normalize` or flattening the input.

* Added ability to set learning rate.

* The sigmoid function in NaturalImage can now be replaced with other functions for testing (In my own experiments, I sometimes found `torch.clamp` created better results).

* Fixed bug with Python 3.6, so that the code works properly on Google Colab.

* Fixed bug with PyTorch 1.7.0. The fix works for earlier versions of PyTorch as well!

### CODE REVIEW

**Comment by ProGamerGov** [comment]
Everything should work correctly now, I think!

**Comment by ProGamerGov** [comment]
@NarineK The diversity objective probably doesn't work right now as there's no support for multiple inputs stacked together currently. But I added it so that @greentfrapp can start working with it's code rather than Lucent's way of doing it.

There's also a second standalone random rescaling class in addition to the rescaling option in RandomAffine, and it uses `torch.nn.functional.interpolate`. I included it as it resizes the input without cropping or creating an unused border, so that we can test how it compares.

I've tested the code myself to make sure things work, and proper tests will be added in a future PR. I haven't added a tutorial notebook yet as using the inception v1 model is pretty much the same as the Torchvision one currently.

**Comment by NarineK** [comment]
Thank you very much for working on this PR, @ProGamerGov! I'll take deeper look into it later this week.

**Comment by ProGamerGov** [comment]
I signed it, but it's still saying that the check was not successful.

**Comment by ProGamerGov** [comment]
I've signed it, but the bot has commented or updated the check yet.

**Comment by ProGamerGov** [comment]
I think it might be important to decide how similar we want transform classes/functions with a Torchvision equivalent to be to the existing Torchvision version.

**Review by NarineK — COMMENTED** [review]
> I think it might be important to decide how similar we want transform classes/functions with a Torchvision equivalent to be to the existing Torchvision version. Obviously users would find it easier if they could use their knowledge from Torchvision, but Captum users may require more flexibility than the current Torchvision way of doing things has.

Sorry for the delay @ProGamerGov! I was in the middle of reviewing this PR them my PTO started. I'll finish my review and post the comments soon.
Having CV transformers similar to torchvision would be great. It would be good to rely on torchvision if there are equivalent transformers available in torchvision package so that we do not have to reimplement the same thing. Which transformers are you exactly talking about ?

File: captum/optim/_core/objectives.py [hunk 1]
```diff
@@ -196,6 +201,45 @@ def loss_function(targets_to_values: ModuleOutputMapping):
...
+def l1(target: nn.Module, constant: float = 0) -> LossFunction:
+    def loss_function(targets_to_values: ModuleOutputMapping):
+        activations = targets_to_values[target]
+        return torch.abs(activations - constant)
+
+    return loss_function
+
+
+def diversity(target: nn.Module) -> LossFunction:
+    def loss_function(targets_to_values: ModuleOutputMapping):
```
[the remark this answers is not in the document]
Comment (ProGamerGov) [reply captum/optim/_core/objectives.py:221]: > Having CV transformers similar to torchvision would be great. It would be good to rely on torchvision if there are equivalent transformers available in torchvision package so that we do not have to reimplement the same thing. Which transformers are you exactly talking about ?

Right now `center_crop` is a function, whereas the main PIL image equivalent in Torchvision is a class. Torchvision also has a RandomAffine class, and each random affine transformation has it's own class for use individually. There's also a functional version of each transform in Torchvision.

I also thought we were trying to avoid requiring Torchvision?

**Comment by ProGamerGov** [comment]
> Having CV transformers similar to torchvision would be great. It would be good to rely on torchvision if there are equivalent transformers available in torchvision package so that we do not have to reimplement the same thing. Which transformers are you exactly talking about ?

@NarineK No worries about the delay! Right now `center_crop` is a function, whereas the main PIL image equivalent in Torchvision is a class. Torchvision also has a RandomAffine class, and each random affine transformation has it's own class for use individually. There's also a functional version of each transform in Torchvision.

I also thought we were trying to avoid requiring Torchvision?

File: captum/optim/_param/image/transform.py [hunk 2]
```diff
@@ -220,7 +314,18 @@ def forward(self, input):
 
 
 class Normalize(nn.Module):
-    def __init__(self, mean, std):
+    """
```
[the remark this answers is not in the document]
Comment (ProGamerGov) [reply captum/optim/_param/image/transform.py:317]: So, I did a quick test and it does seem like we can just use `torchvision.transforms.Normalize` for this transform as it supports tensors and is autograd compatible.

File: captum/optim/_core/objectives.py [hunk 3]
```diff
@@ -196,6 +201,45 @@ def loss_function(targets_to_values: ModuleOutputMapping):
     return loss_function
 
 
+def deepdream(target: nn.Module) -> LossFunction:
+    def loss_function(targets_to_values: ModuleOutputMapping):
+        activations = targets_to_values[target]
+        return activations ** 2
```
[the remark this answers is not in the document]
Comment (ProGamerGov) [reply captum/optim/_core/objectives.py:207]: I squared the activation as that's how Lucid does it, and I've reached out for clarification as to why they do it. I also performed a quick test to compare with and without squaring the activtions.

Without squaring the activations:

output_no_pow

With squaring the activations:

output_pow

[the remark this answers is not in the document]
Comment (ProGamerGov) [reply captum/optim/_core/objectives.py:207]: The shape of the activations shouldn't matter for this objective as it's just maximizing the activations at the target layer.

File: captum/optim/_param/image/transform.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ProGamerGov) [reply captum/optim/_param/image/transform.py:317]: It looks like we may be able to use Torchvision transforms for pretty much all of the transforms, as they are autograd compatible.

File: captum/optim/_core/objectives.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ProGamerGov) [reply captum/optim/_core/objectives.py:240]: Okay, I've added a short description, a citation, and a link for the Distill article that it comes from!

File: captum/optim/_core/objectives.py [hunk 4]
```diff
@@ -196,6 +201,45 @@ def loss_function(targets_to_values: ModuleOutputMapping):
     return loss_function
 
 
+def deepdream(target: nn.Module) -> LossFunction:
+    def loss_function(targets_to_values: ModuleOutputMapping):
```
[the remark this answers is not in the document]
Comment (ProGamerGov) [reply captum/optim/_core/objectives.py:214]: That sounds like a really good idea! Though, let's put in a different PR.

**Comment by ProGamerGov** [comment]
> Sorry for the delay @ProGamerGov! I was in the middle of reviewing this PR them my PTO started. I'll finish my review and post the comments soon.
> Having CV transformers similar to torchvision would be great. It would be good to rely on torchvision if there are equivalent transformers available in torchvision package so that we do not have to reimplement the same thing. Which transformers are you exactly talking about ?

No worries! It actually looks like we may be able to just use Torchvision transforms for pretty much all of the transforms, as they are autograd compatible. I had thought they still only supported PIL images and didn't realize that they were updated at some point.

File: captum/optim/_core/objectives.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ProGamerGov) [reply captum/optim/_core/objectives.py]: I changed the power value to be a user specified value for now.

**Comment by ProGamerGov** [comment]
I have removed RandomAffine in favour of Torchvision's RandomAffine. I kept RandomScale as there isn't an exact Torchvision equivalent.

File: captum/optim/_param/image/transform.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ProGamerGov) [reply captum/optim/_param/image/transform.py]: Okay, I've replaced the normalize function with the torchvision one. I've also replaced some of the other transforms where applicable!

File: captum/optim/_core/objectives.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ProGamerGov) [reply captum/optim/_core/objectives.py]: I think for the moment we can leave it as a user controllable value for power, as I'm sure there's a good reason Ludwig and his team did it that way. But I'll continue looking into it.

[the remark this answers is not in the document]
Comment (ProGamerGov) [reply captum/optim/_core/objectives.py]: @greentfrapp just helped figure out what I think is probably the actual reason. Squaring the activations removes negative values.

[the remark this answers is not in the document]
Comment (ProGamerGov) [reply captum/optim/_core/objectives.py]: @NarineK Okay, so I spoke with Chris about it and if the square is removed, it'd be doing a direciton visualization of [1,1,1,....] instead of DeepDream. 

> DeepDream tries to increase the values of neurons proportional to the amount they are presently active which is equivalent to maximizing the sum of the squares.

File: captum/optim/_param/image/images.py [hunk 5]
```diff
@@ -136,9 +137,9 @@ class ToRGB(nn.Module):
     def klt_transform():
```
Comment (NarineK) [inline captum/optim/_param/image/images.py:137]: Isn't `ToRBG` also a transformation and we could move it to `/image/transform.py` ?

File: captum/optim/_param/image/transform.py [hunk 6]
```diff
@@ -1,11 +1,9 @@
-import logging
 import math
 import numbers
 
 import torch
 import torch.nn as nn
 import torch.nn.functional as F
-from kornia.geometry.transform import rotate, scale, shear, translate
 
 device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
```
Comment (NarineK) [inline captum/optim/_param/image/transform.py:8]: Do we need this ?

File: captum/optim/_param/image/images.py [hunk 7]
```diff
@@ -1,8 +1,8 @@
-import imageio
 import matplotlib.pyplot as plt
 import numpy as np
 import torch
 import torch.nn as nn
+from PIL import Image
```
Comment (NarineK) [inline captum/optim/_param/image/images.py:5]: In terms of PIL, I think that our standard install setup might not have PIL package support. We can have this import in try-expect and ask the users to install it. See the example: https://github.com/pytorch/captum/pull/468/files#diff-0399fdfb934ad0d1e07d13305904829c89866b3c4adce935da861aff725d03fcR46

**Review by NarineK — APPROVED** [review]
Thank you for working on this PR @ProGamerGov! I left couple minor comments. 
In terms of PIL, we might need to put it in try-except and ask users to install it in case it is not installed. I added an example.

File: captum/optim/_param/image/transform.py (hunk 6 above)
Comment (ProGamerGov) [reply captum/optim/_param/image/transform.py:8]: Nope, we no longer need Kornia for anything! The rotate and shear transforms from Torchvision can be used instead, and thus I removed some of the Kornia replacement code I wrote. The translation and scale transforms technically exist in Torchvision as well under RandomAffine, but they are not setup ideally for what we need so I kept those versions.

File: captum/optim/_param/image/images.py (hunk 5 above)
Comment (ProGamerGov) [reply captum/optim/_param/image/images.py]: Yeah I guess it is! I've moved it to `transform.py`.

File: captum/optim/_param/image/transform.py (hunk 6 above)
Comment (ProGamerGov) [reply captum/optim/_param/image/transform.py:8]: The two example notebooks showcase examples of using both Torchvision & Captum.optim transforms at the same time.

File: captum/optim/_param/image/images.py (hunk 7 above)
Comment (ProGamerGov) [reply captum/optim/_param/image/images.py]: Alright, I've added a try-expect block for alerting users that Pillow/PIL is required if it's not already installed!

File: captum/optim/_core/objectives.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (NarineK) [reply captum/optim/_core/objectives.py]: Thank you for clarifying this, @ProGamerGov! Perhaps, we can add that explanation in the code so that we don't forget it.

File: captum/optim/_param/image/transform.py (hunk 6 above)
Comment (NarineK) [reply captum/optim/_param/image/transform.py:8]: Awesome! Thank you for do that!

File: captum/optim/_core/objectives.py [hunk 8]
```diff
@@ -1,258 +1,349 @@
...
+    def loss(self, targets_to_values):
+        assert len(self.targets) == 1
+        target = self.targets[0]
+        target_value = targets_to_values[target]
+        loss_value = self.loss_function(target_value)
+        self.history.append(loss_value.sum().cpu().detach().numpy().squeeze().item())
+        return loss_value
+
+
+class MultiObjective(Objective):
```
Comment (NarineK) [inline captum/optim/_core/objectives.py:319]: nit: Do we want to call this: `MultiTargetObjective` to be consistent of `SingleTargetObjective ` ?

**Review by NarineK — APPROVED** [review]

File: captum/optim/_param/image/images.py (hunk 7 above)
Comment (NarineK) [reply captum/optim/_param/image/images.py]: Thank you!

File: captum/optim/_core/objectives.py (hunk 8 above)
Comment (ProGamerGov) [reply captum/optim/_core/objectives.py]: Yes we probably should call it `MultiTargetObjective` instead of `MultiObjective`. I've changed the name.

**Comment by ProGamerGov** [comment]
@NarineK If there's nothing else that needs to be changed currently, then I think this pull request should be ready for merging?

**Comment by NarineK** [comment]
> @NarineK If there's nothing else that needs to be changed currently, then I think this pull request should be ready for merging?

Thank you for making the changes @ProGamerGov ! Yes, we can merge this PR. Last time the system didn't allow SK to merge. I think they allow only the owners or employees to merge. If it doesn't allow you to merge it, I can press the button.

**Comment by ProGamerGov** [comment]
@NarineK Yeah, it looks like only those with write access can merge PRs. So, you'll have to merge it for me.

**Comment by NarineK** [comment]
Merged, thank you @ProGamerGov !
